rfd/0177-upgrade-js-package-manager.md

```diff
@@ -0,0 +1,183 @@
+Here pnpm appears to be the winner. It supports installing dependencies as 
+linked by default, which makes it the most battle-tested implementation.
+The speed is also great, it's consistently one of the fastest tools in 
+[benchmarks](https://p.datadoghq.eu/sb/d2wdprp9uki7gfks-c562c42f4dfd0ade4885690fa719c818?fromUser=false&refresh_mode=sliding&tpl_var_npm%5B0%5D=%2A&tpl_var_pnpm%5B0%5D=%2A&tpl_var_yarn-classic%5B0%5D=%2A&tpl_var_yarn-modern%5B0%5D=%2A&tpl_var_yarn-nm%5B0%5D=%2A&tpl_var_yarn-pnpm%5B0%5D=no&from_ts=1711447153431&to_ts=1719223153431&live=true).
+
+The npm support for `linked` is really interesting, but as long as it is 
+experimental, we should not use it in a production environment.
+
+When it comes to Yarn, the main advantage is that we already use it, so we could 
+still type `yarn start-teleport`. 
+However, if we are going to use pnpm installation method, why not switch to pnpm 
+entirely? 
+Yarn also shows a warning `The pnpm linker doesn't support providing different 
+versions to workspaces' peer dependencies` during the installation, but I'm not 
+sure what significance this has.
+
+## Migration process
+### Establishing correct relationships between workspaces
+After switching the package manager to pnpm, the Web UI/Connect apps won’t run, 
+due to TS/JS code in the workspaces not being able to resolve their imports 
+(like `react`).
+Why? This is a continuation of the hoisting problem — our workspaces rely 
+on dependencies of other workspaces, without declaring them in their 
+`package.json`.
+It worked in Yarn 1.x, since adding a dependency to a one workspace made it 
+available in any other workspace (because of hoisting to the root 
+`node_modules`).
+We can see this in `packages/build`: the `@opentelemetry/*` dependencies are 
+declared there but used only in `packages/teleport`.
+
+This contradicts the idea of workspaces: they are designed to encapsulate 
+project dependencies and code. Sharing dependencies between unrelated workspaces 
+breaks this encapsulation, leading to potential conflicts and contamination 
+where changes in one workspace inadvertently affect another.
+
+To allow the projects to compile again, the dependencies need to be moved to the
+correct workspaces.
+I see three scenarios for how they should be moved over:
+1. Dependencies that are used only in a single workspace should be kept there.
+Example: `electron-builder` that builds Connect - it’s declared in 
+`packages/build`, but it should be in `packages/teleterm`.
+2. Project-wide dependencies imported everywhere, should be kept in the root 
+`package.json`.
+Example: `react` - declared in `build/package.json`, but imported in every other 
+workspace.
+3. Project-wide tools used in the root `package.json` should be kept in 
+`build/package.json`, if possible. 
+Example: `jest` or `storybook` can be installed in `build/package.json` 
+which will expose them through scripts to the root `package.json`.
```

Comment:
What's the purpose behind a separate build package then?

I believe the package structure that we have today comes from Gravity and Teleport which had to share common deps. Gravity got discontinued, but the package structure remains, because we now have Connect which also needs the same deps as Teleport.

The other question is…

## Do we need workspaces at all?

Where are workspaces useful in the first place? From my experience, they're useful when inside a single repo you work on multiple individual packages that are consumed separately, and those packages depend on a bunch of smaller packages that are in the repo. "Consumed separately" might mean they're published as separate npm packages. Or there's multiple apps and separate CI jobs want to pull in only deps required by a certain app.

When working on those separate packages, you don't want to be limited by the regular npm package limitations. I.e., if you want to update `<Button>`, it's best if you could just update one file and both `teleterm` and `teleport` picked up the changes immediately. It'd be bad if you had to release `design@1.0.2` first and then update it in both `teleterm` and `teleport`.

Perhaps also those separate packages need separate deps, or separate versions of deps.

Quoting from [Yarn docs](https://yarnpkg.com/features/workspaces#when-should-i-use-workspaces):

> Workspaces shine in many situations, with the main ones being:
>
> *    When a project has a core package surrounded with various add-ons; this is for example the case for Babel.
>
> *    When you need to publish multiple packages and want to avoid your contributors having to open PRs on many separate repositories whenever they want to make a change. This is for example the case for Jest.
>
> *    Or when projects want to keep strict boundaries within their code and avoid becoming an entangled monolith. This is for example the case for Yarn itself, or many enterprise codebases.

The last point might sound as if it applied to us.

## How do we use workspaces at the moment?

Currently, we do none of that. We always install all deps, we always clone the whole repo. There's pretty much no dep separation. Almost all dep management is done by Vite which just pulls relevant deps into the final bundle.

On the contrary – we want both projects (Teleport and Connect) to use the same versions of deps so that it's easy to reuse code between them. So the way pnpm resolves deps is beneficial here – we want the same deps always, everywhere.

The way we use workspaces feels similar to how you'd use regular CommonJS modules to share code. `<Button>` is used in multiple places, so you place it in `button/index.js` and use it in multiple places. `vite` is used by both projects, so you place it in `build/package.json` and use it in multiple places.

But as you mentioned, that's not how packages work in. The closest we are to using packages the "correct" way is `shared` and `design` maybe, where I don't really care how `<Button>` is implemented or what deps it needs, I just want to use it in my component. So I import `Button` from `design` and yarn handles `design` deps during installation.

## Could we live without workspaces?

Connect needs a separate package.json so that electron-builder can pick correct deps. This is the one place where dep separation actually matters. Otherwise each .dmg with Connect would ship with JS deps it doesn't use at all.

Could we build Connect without workspaces though? What if it just continued to be a separate folder? Could we survive not using workspaces or are there are benefits to them that I failed to notice?

Reply:
It seems that if Connect is supposed to stay consumed as a separate entity by electron-builder, it needs to behave as a real package. As in, it needs to have its own package.json and the resolved packages need to be accounted for in some lockfile.

So maybe the structure proposed by Grzegorz is fine? A separate `build` package certainly helps with encapsulating which deps are needed for dev purposes only.

Reply:
>Could we build Connect without workspaces though? What if it just continued to be a separate folder?

When it comes to `electron-builder`, we could probably use the two package.json structure https://www.electron.build/tutorials/two-package-structure.html?
So in the `teleterm/package.json` we would still keep the dependencies that should be bundled.
The drawback of it is that we would have to add `postinstall` script to trigger installing them:
>To ensure your dependencies are always updated based on both files, simply add "postinstall": "electron-builder install-app-deps" to your development package.json. This will basically automatically trigger an npm install within your app directory so you don’t have to do this work every time you install/update your dependencies.

But tbh, I'm not a huge fun of this structure, it feels more complicated than the workspace option.  

When it comes to workspaces in general, I think the biggest benefit for us is this one:
>Or when projects want to keep strict boundaries within their code and avoid becoming an entangled monolith. This is for example the case for Yarn itself, or many enterprise codebases.

Currently we break these boundaries by importing from `teleport` to `shared` or to `teleterm`.
However, while working on the RFD, I realized that we could easily enforce strict rules between the workspaces by getting rid of TS aliases and importing the code directly from the packages, for example `@gravitational/shared/AccessRequests` instead of `shared/AccessRequests`. Then we would have to set `hoist-workspace-packages=false` in the config to prevent linking all workspaces to the root `node_modules`.

After that, a given workspace would be able to import code only from the workspaces that it declares in its `package.json`.

Maybe it's possible to achieve a similar result with a linter rule? But for sure enforcing this at the package level provides better DX than during lining.

Reply:
>What's the purpose behind a separate build package then?

I wasn't 100% sure if we should keep it, after we fully migrate to Vite, we will only have `jest` and `vite` configs there, plus the dependencies. Mainly, I wanted to not make too many changes in this migration.

However, maybe it will be better to keep the dev dependencies in the root package.json (some of them will need to be moved there anyway and packages/build will shrink even more).

Reply:
I'm not sure how much of it applies to pnpm, but when I was reading about Plug'n'Play, I found this:

> Yarn prevents ghost dependencies in the packages your project depends on, but also in your own code itself - this is to decrease the chances that a package would work on your development machine but break once published.
> 
> It however has a side effect when it comes to bins. If you have typescript listed at the root of your project, the tsc binary will be available in the root package but only in the root project. In other words, any workspace using the tsc binary in its scripts will need to declare it in its dependencies.
> 
> A good recommendation to avoid this kind of issue is to have a "tooling" workspace to contain your infrastructure tools and scripts, and have all other workspaces depend on it.

https://yarnpkg.com/features/pnp#shared-binaries

Reply:
I tested this scenario in `pnpm` and adding typescript to the root `package.json` makes `tsc` available in a child workspace, so we wouldn't have this problem.

But I think I will still try to keep the tooling in `packages/build` and see how it works.